### PR TITLE
GODRIVER-2446 Update Fixable Dep Vulnerabilities, Identify Un-Fixable Dep Vulnerabilities

### PR DESCRIPTION
**Pull request by prestonvasquez**
1. golang.org/x/net CVE-2021-44716 :white_check_mark:
2. golang.org/x/text CVE-2020-14040 :white_check_mark:
3. golang.org/x/text CVE-2021-38561 :white_check_mark:
4. golang.org/x/crypto CVE-2020-9283, CVE-2020-29652, CVE-2022-27191 :white_check_mark:
5. gopkg.in/yaml.v3 CVE-2022-28948

### `golang.org/x/net` 

[GODRIVER-2448](https://jira.mongodb.org/browse/GODRIVER-2448)

[CVE-2021-44716](https://www.cve.org/CVERecord?id=CVE-2021-44716)

We have two packages that use the net library:

1. `golang.org/x/crypto`
2. `golang.org/x/tools`

Updating both to latest respectively puts the dependency graphs at

```
golang.org/x/crypto@v0.0.0-20220622213112-05595931fe9d golang.org/x/net@v0.0.0-20211112202133-69e39bad7dc2

golang.org/x/crypto@v0.0.0-20210921155107-089bfa567519 golang.org/x/net@v0.0.0-20210226172049-e18ecbb05110
```

and

```
golang.org/x/tools@v0.1.11 golang.org/x/net@v0.0.0-20211015210444-4f30a5c0130f

golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e golang.org/x/net@v0.0.0-20190620200207-3b0461eec859
```

[According to Snyk](https://app.snyk.io/org/mongodb-default/project/d1fd28ee-07d4-4de7-88ed-e20dc42a307c#issue-SNYK-GOLANG-GOLANGORGXNETHTTP2-2313688) the DOS vulnerability for `golang.org/x/net` is fixed in

```
golang.org/x/net@0.0.0-20211209124913-491a49abca63
```

**which means that neither of our latest dependencies resolves this problem**, as our earliest version for the net library is

```
golang.org/x/net@v0.0.0-20190620200207-3b0461eec859
```

The root of this 2019 version is that `golang.org/x/tools` [imports an older version of itself](https://cs.opensource.google/go/x/tools/+/master:go.sum;l=29). This peculiarity results in the following dependency graph:

```
golang.org/x/tools@v0.1.11 golang.org/x/mod@v0.6.0-dev.0.20220419223038-86c51ed26bb4
golang.org/x/mod@v0.6.0-dev.0.20220419223038-86c51ed26bb4 golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e
golang.org/x/tools@v0.0.0-20191119224855-298f0cb1881e golang.org/x/net@v0.0.0-20190620200207-3b0461eec859
golang.org/x/net@v0.0.0-20190620200207-3b0461eec859 golang.org/x/crypto@v0.0.0-20190308221718-c2843e01d9a2
```

The only place in our repo that uses x/tools is [docbuilder](https://github.com/mongodb/mongo-go-driver/blob/master/cmd/docbuilder/main.go). The goal of this package was to create a binary that generates HTML for package redirects, defined in [GODRIVER-120](https://jira.mongodb.org/browse/GODRIVER-120). Since the go team no longer maintains this practice, we can deprecate the docbuilder and do only top-level redirects as requested in ~~[TSOPS-4909](https://jira.mongodb.org/browse/TSOPS-4909)~~ ~~[HELP-35307](https://jira.mongodb.org/browse/HELP-35307)~~ [ITSYSENG-5757](https://jira.mongodb.org/browse/ITSYSENG-5757).

The x/crypto package still uses a vulnerable version of the net package:

```
golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2
```

~~To help get closer to resolving this, I've submitted an issue via the [Go Security Policy](https://go.dev/security/policy):~~

> ~~Hi,~~
>
> ~~It appears that the x/crypto package still uses a version of golang.org/x/net that is still subject to the CVE-2021-44716 vulnerability. Surprisingly there is no issue reported for this at the moment.~~

However, per the go team:

> As far as I am aware x/net/http2 is not used anywhere in x/crypto, the only x/net package used is x/net/idna, which is unaffected by this security issue.

Given this issue is specifically concerning the use of x/net/http2, this vulnerability is resolved. It's also worth noting that the dependency update given by Snyk is specific to the x/tools package.

### `golang.org/x/text`

[GODRIVER-2447](https://jira.mongodb.org/browse/GODRIVER-2447)

#### CVE-2020-14040

This is a DOS vulnerability that is fixed in `golang.org/x/text@0.3.3`. The earliest version of the `golang.org/x/text` package in the Go Driver is `golang.org/x/text@v0.3.6`, which is a dependency of `golang.org/x/crypto`:

```
golang.org/x/crypto@v0.0.0-20220622213112-05595931fe9d golang.org/x/text@v0.3.6
```

#### CVE-2021-38561

[This no longer appears to be an issue on cve.org](https://www.cve.org/CVERecord?id=CVE-2021-38561). ~~However, it is still a vulnarbility on Snyk that is fixed as of `golang.org/x/text@v0.3.7`. In the Go Driver, this vulnerability is introduced by `golang.org/x/crypto`: https://cs.opensource.google/go/x/crypto/+/master:go.mod;l=11 and~~ the question of CVE-2021-38561's effects on the x/crypto package is currently an [open issue](https://github.com/golang/go/issues/51575) with their team. This is not, however, an issue reported by Snyk AFAIK.

~~I've created [GODRIVER-2447|https://jira.mongodb.org/browse/GODRIVER-2477] to resolve this whenever the Go team's issue is closed.~~

This is resolved by updating `github.com/xdg-go/stringprep`, per the dependency graph:

```
github.com/xdg-go/stringprep@v1.0.3 golang.org/x/text@v0.3.7
```

### `golang.org/x/crypto`

[GODRIVER-2446](https://jira.mongodb.org/browse/GODRIVER-2446)

[CVE-2020-9283](https://www.cve.org/CVERecord?id=CVE-2020-9283)
[CVE-2020-29652](https://www.cve.org/CVERecord?id=CVE-2020-29652)
[CVE-2022-27191](https://www.cve.org/CVERecord?id=CVE-2022-27191)

All three of these issues are resolved by updating `golang.org/x/crypto` to latest.

### `gopkg.in/yaml.v3`

[GODRIVER-2448](https://jira.mongodb.org/browse/GODRIVER-2448)

[CVE-2022-28948](https://www.cve.org/CVERecord?id=CVE-2022-28948)

This is resolved by upgrading gopkg.in/yaml.v3 to version 3.0.0 or higher. I believe replacing the effected version in the module will resolve this issue: https://go.dev/ref/mod#:~:text=A%20replace%20directive%20replaces%20the,a%20platform%2Dspecific%20file%20path.

### CODE REVIEW

**Review by matthewdale — APPROVED** [review]
Looks good 👍 . Thanks for the thorough description of the updates!

File: go.mod [hunk 1]
```diff
@@ -7,6 +7,11 @@ retract (
 	[v1.6.0, v1.6.1] // Contains data race bug in background connection establishment.
 )
 
+//gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c introduced through github.com/stretchr/testify@v1.6.1 are
+// vulnerable to Denial of Service (DoS) via the Unmarshal function, which causes the program to crash when attempting
+// to deserialize invalid input. https://www.cve.org/CVERecord?id=CVE-2022-28948
+replace gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c => gopkg.in/yaml.v3 v3.0.1
```
Comment (prestonvasquez) [inline go.mod:13]: @matthewdale as you've noted, updating to a version of testify that resolves CVE-2022-28948 is not something we can do without breaking support for Go v10. Using [replace](https://go.dev/ref/mod#:~:text=A%20replace%20directive%20replaces%20the,a%20platform%2Dspecific%20file%20path.) updates the vulnerable yaml.v3 version to 3.0.1, which removes it from go.sum. However, the testify package still points to `v3.0.0-20200313102051-9f266ea9e77c` in the dependency graph:

```
github.com/stretchr/testify@v1.6.1 gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c
```

File: go.mod [hunk 2]
```diff
@@ -7,6 +7,11 @@ retract (
 	[v1.6.0, v1.6.1] // Contains data race bug in background connection establishment.
 )
 
+// gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c introduced through github.com/stretchr/testify@v1.6.1 are
+// vulnerable to Denial of Service (DoS) via the Unmarshal function, which causes the program to crash when attempting
+// to deserialize invalid input. https://www.cve.org/CVERecord?id=CVE-2022-28948
+replace gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c => gopkg.in/yaml.v3 v3.0.1
```
Comment (prestonvasquez) [inline go.mod:13]: @matthewdale as you've noted, updating to a version of testify that resolves CVE-2022-28948 is not something we can do without breaking support for Go v10. Using [replace](https://go.dev/ref/mod#:~:text=A%20replace%20directive%20replaces%20the,a%20platform%2Dspecific%20file%20path.) updates the vulnerable yaml.v3 version to 3.0.1, which removes it from go.sum. However, the testify package still points to `v3.0.0-20200313102051-9f266ea9e77c` in the dependency graph:

```
github.com/stretchr/testify@v1.6.1 gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c
```

Comment (matthewdale) [reply go.mod:13]: I think that's fine. The only version of gopkg.in/yaml.v3 in `go.sum` seems to be v3.0.1, suggesting that `replace` directive is preventing it from being part of the final dependency closure.

**Comment by benjirewis** [comment]
Wow thanks for the super thorough description! That move with the `replace` directive is really smart... As for [CVE--2021-44716](https://github.com/advisories/GHSA-vc3p-29h2-gpcp), is there no remaining Snyk alert from the `x/crypto` package relying on a vulnerable version of `x/net`?

**Review by benjirewis — APPROVED** [review]
Gotcha. If the Go team says we're not using the vulnerable code with our current setup and Snyk is not complaining I think that's fine w me. Thanks!

**Pull request merged**